Add a constructor for mixture distributions

marcoct opened the pull request:
This PR adds a macro for defining a mixture distribution family for a base distribution family.

Example usage:
```julia
mixture_of_normals = @mixturedist normal (0, 0) Float64

@gen function foo()
     x ~ mixture_of_normals([0.4, 0.5], [-1.0, 1.0], [1.0, 1.0])
end
```
The first argument to the macro is the base distribution, which should be a Gen.Distribution.
The second argument defines whether the arguments to the base distribution are scalars or `Array`s and if they are `Array`s, what their dimension is.
The third argument defines the output type, which should match the output type of the base distribution.

Currently, only distributions with all-scalar arguments are supported (e.g. mixtures of multivariate normal is not supported).

marcoct committed:
initial mixturedist, with tests

marcoct committed:
fix bugs, add to docs, test binom mixture

marcoct committed:
add missing test line

ztangent commented:
Yay! It seems like it'd be nice to also support mixtures of different base distributions? e.g. normals + continuous uniforms?

Also, I wonder if a shorter `@mixture` macro would be enough -- would that clash with anything?

ztangent commented:
It looks like Distributions.jl also has a `MixtureModel` structure: https://juliastats.org/Distributions.jl/stable/mixture/

marcoct committed:
Support distributions with non-scalar arguments

marcoct committed:
Add heterogeneous mixtures

marcoct committed:
Update docs

marcoct commented:
I added support for heterogeneous mixtures formed by a fixed list of base distributions, so now there are two macros, one for defining fixed-length heterogeneous mixtures, and the other for defining variable-length homogeneous mixtures.

I also added support for base distributions that take non-scalar arguments, and this functionality is tested for a homogeneous mixture of multivariate normal distributions.

marcoct committed:
add missing finite diff helper

marcoct committed:
improve code readability

alex-lew commented on src/modeling_library/mixture.jl:
```diff
@@ -1,33 +1,47 @@
 Example:
 
     mixture_of_normals = @hom_mixture Float64 normal (0, 0)
 """
```
Can we update these docs to also describe the usage of the resulting mixture distribution? (It's not obvious to me that the calling pattern is `dist(weights, arg1_for_all_components, arg2_for_all_components, ...)`.)

marcoct replied:
> * Why do we need this to be a macro, rather than a 'distribution combinator'? Is it that this keeps more avenues open for performance engineering in the future?
> *

That was ostensibly the idea. But I'll see what a non-macro version looks like.

alex-lew commented on src/modeling_library/mixture.jl:
```diff
@@ -37,131 +51,145 @@ macro hom_mixture(output_type, base_dist, arg_dims)
 
-const mixed_syntax_err_msg ="Expected syntax of the form @mixed Float64 (normal, 2) (uniform, 2)"
-const mixed_weight_vector_err_msg = "length of weight vector must match the number of components"
+const MIXED_SYNTAX_ERR_MSG ="Expected syntax of the form @mixed Float64 (normal, 2) (uniform, 2)"
```
Should this say `@het_mixture` rather than `@mixed`?

alex-lew commented on src/modeling_library/mixture.jl:
```diff
@@ -0,0 +1,283 @@
+Example:
+
+    uniform_beta_mixture = @het_mixture Float64 (uniform, 2) (beta, 2)
+"""
```
Can we clarify here too that the resulting calling pattern is `dist(weights, args_for_dist_1..., args_for_dist_2...)`?

alex-lew commented:
@marcoct This is great! I left a couple comments, but overall I'm happy to merge.

I do have some questions for the longer term:

1. Why do we need this to be a macro, rather than a 'distribution combinator'? Is it that this keeps more avenues open for performance engineering in the future?

2. I wonder if the interfaces for `hom_mixture` and `het_mixture` can be made more similar -- e.g., you could imagine a unified calling pattern for the created distributions (maybe `dist(weights, component_1_args, component_2_args, ...)`). If each set of `args` is in its own tuple / vector, then you wouldn't even need the user to supply the information about number of arguments (or argument dimension, if I understand correctly how that info is currently being used).

marcoct commented:
> I wonder if the interfaces for hom_mixture and het_mixture can be made more similar -- e.g., you could imagine a unified calling pattern for the created distributions (maybe dist(weights, component_1_args, component_2_args, ...)). If each set of args is in its own tuple / vector, then you wouldn't even need the user to supply the information about number of arguments (or argument dimension, if I understand correctly how that info is currently being used).

I decided to vectorize arguments in the homogeneous case so that ReverseDiff.jl-based AD would work out of the box (it works with arrays, but not with tuples). (Also, capturing the homogeneity in in the interface might also be useful for provisioning for performance optimizations.)

marcoct commented:
I implemented a non-macro version of the homogeneous mixture case, and there is a 1-3x slow down for the test cases, maybe due to increased dependency on splatting. But the code is of course substantially easier to read.

alex-lew commented:
> I implemented a non-macro version of the homogeneous mixture case, and there is a 1-3x slow down for the test cases, maybe due to increased dependency on splatting. But the code is of course substantially easier to read.

@marcoct Oh, that reminds me — one worry about the `arg[:, :, j]` thing is that I think it performs a copy?

marcoct commented:
I also implemented a non-macro version of the heterogenous case, which resulted in a 2x slowdown.

marcoct committed:
add non-macro implementations

marcoct committed:
Use the non-macro implementations

marcoct committed:
Use non-macro versions in docs

marcoct committed:
expand docs for homogeneous mixture

marcoct committed:
Fix error message

marcoct committed:
Rearrange in docs

marcoct committed:
add docs for heterogeneous mixture

marcoct committed:
Fix docs

marcoct committed:
Fix capitalization in docs

marcoct committed:
improve docs formatting

marcoct committed:
tweak docs

marcoct committed:
add a link in the docs
